### PR TITLE
Top k udaf

### PR DESCRIPTION
**Pull request by hjafarpour**
This PR adds `TopK` and `TopKDistinct` aggregate functions. The result of these functions is an array with size `k`. Currently we can accept BIGINT, DOUBLE and VARCHAR as input parameters. 
The result is an array with descending values. TopKDistinct will not have duplicate values in the result.
This PR is for https://github.com/confluentinc/ksql/issues/554.

### CODE REVIEW

**Comment by hjafarpour** [comment]
retest this please.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+  @Override
+  public Merger<String, Double[]> getMerger() {
+    // TODO: For now we just use a simple algorithm. Maybe try finding a faster algorithm later
+    return (aggKey, aggOne, aggTwo) -> {
+      int nullIndex1 = ArrayUtil.getNullIndex(aggOne) == -1? tkVal: ArrayUtil.getNullIndex(aggOne);
+      int nullIndex2 = ArrayUtil.getNullIndex(aggTwo) == -1? tkVal: ArrayUtil.getNullIndex(aggTwo);
+      Double[] tempMergeTopkArray = new Double[nullIndex1 + nullIndex2];
+
+      for (int i = 0; i < nullIndex1; i++) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java]: `System.arraycopy(..)` ?  and below

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    super(argIndexInValue, new Double[tkVal], SchemaBuilder.array(Schema.FLOAT64_SCHEMA).build(),
+          Arrays.asList(Schema.FLOAT64_SCHEMA),
+          "TOPK", DoubleTopkKudaf.class);
+    this.tkVal = tkVal;
+    this.topkArray = new Double[tkVal];
+    this.tempTopkArray = new Double[tkVal + 1];
+  }
+
+  @Override
+  public Double[] aggregate(Double currentVal, Double[] currentAggVal) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java]: Tests for the values this returns under different conditions? There is one test but it only asserts after multiple values have been added to the array. What about when the first value is added?

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    return Arrays.copyOf(tempTopkArray, tkVal);
+  }
+
+  @Override
+  public Merger<String, Double[]> getMerger() {
+    // TODO: For now we just use a simple algorithm. Maybe try finding a faster algorithm later
+    return (aggKey, aggOne, aggTwo) -> {
+      int nullIndex1 = ArrayUtil.getNullIndex(aggOne) == -1? tkVal: ArrayUtil.getNullIndex(aggOne);
+      int nullIndex2 = ArrayUtil.getNullIndex(aggTwo) == -1? tkVal: ArrayUtil.getNullIndex(aggTwo);
+      Double[] tempMergeTopkArray = new Double[nullIndex1 + nullIndex2];
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java]: Can we just always make the size of this array to be `tkVal * 2` and then i think we don't need the 
```
if (tempMergeTopkArray.length < tkVal) {
        tempMergeTopkArray = ArrayUtil.padWithNull(Double.class, tempMergeTopkArray, tkVal);
        return tempMergeTopkArray;
      }
```
below??

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  public KsqlAggregateFunction<Double, Double[]> getInstance(Map<String, Integer> expressionNames,
+                                                             List<Expression> functionArguments) {
+    if (functionArguments.size() != 2) {
+      throw new KsqlException(String.format("Invalid parameter count. Need 2 args, got %d arg(s)"
+                                            + ".", functionArguments.size()));
+    }
+    int udafIndex = expressionNames.get(functionArguments.get(0).toString());
+    Integer tkValFromArg = Integer.parseInt(functionArguments.get(1).toString());
+    DoubleTopkKudaf doubleTopkKudaf = new DoubleTopkKudaf(udafIndex, tkValFromArg);
+    return doubleTopkKudaf;
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java]: nit: inline

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/LongTopkKudaf.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+          Arrays.asList(Schema.INT64_SCHEMA),
+          "TOPK", LongTopkKudaf.class);
+    this.tkVal = tkVal;
+    this.topkArray = new Long[tkVal];
+    this.tempTopkArray = new Long[tkVal + 1];
+  }
+
+  @Override
+  public Long[] aggregate(Long currentVal, Long[] currentAggVal) {
+    if (currentVal == null) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/LongTopkKudaf.java]: This method and the next look identical to the one in `DoubleTopKudaf` refactor the common parts so the algorithm is only expressed once and can be changed once in the future

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/StringTopkKudaf.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    super(argIndexInValue, new String[tkVal], SchemaBuilder.array(Schema.STRING_SCHEMA).build(),
+          Arrays.asList(Schema.STRING_SCHEMA),
+          "TOPK", StringTopkKudaf.class);
+    this.tkVal = tkVal;
+    this.topkArray = new String[tkVal];
+    this.tempTopkArray = new String[tkVal + 1];
+  }
+
+  @Override
+  public String[] aggregate(String currentVal, String[] currentAggVal) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/StringTopkKudaf.java]: As above

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/DoubleTopkDistinctKudaf.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.Map;
+
+import io.confluent.ksql.function.KsqlAggregateFunction;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.util.ArrayUtil;
+import io.confluent.ksql.util.KsqlException;
+
+public class DoubleTopkDistinctKudaf extends KsqlAggregateFunction<Double, Double[]> {
+
+  Integer tkVal;
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/DoubleTopkDistinctKudaf.java]: same as in other one. fields should be private, `topKarray` is not needed. `aggregate` and `merger` functions do the same thing across all implementions.

**Comment by satybald** [comment]
Fixes #554?

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import io.confluent.ksql.util.KsqlException;
+
+
+public class  DoubleTopkKudaf extends KsqlAggregateFunction<Double, Double[]> {
+
+  Integer tkVal;
+  Double[] topkArray;
+  Double[] tempTopkArray;
+
+  public DoubleTopkKudaf(Integer argIndexInValue, Integer tkVal) {
```
[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java]: Made this and all other ones package-private.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java (hunk 2 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudaf.java]: Added tests for this and the other ones.

**Comment by hjafarpour** [comment]
@satybald Yes, this is for https://github.com/confluentinc/ksql/issues/554.

File: ksql-engine/src/main/java/io/confluent/ksql/util/ArrayUtil.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    for (int i = array.length; i < finalLength; i++) {
+      paddedArray[i] = null;
+    }
+    return paddedArray;
+  }
+
+  public static <T> boolean containsValue(T value, T[] array) {
+    for (T v: array) {
+      if (v != null && v.toString().equals(value.toString())) {
```
Comment (satybald) [inline ksql-engine/src/main/java/io/confluent/ksql/util/ArrayUtil.java]: what's the need to convert valuet to string? Why not directly compare value with `equals`?

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudafTest.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+    assertThat("Invalid results.", doubleTopkKudaf.getMerger().apply("key", array1, array2), equalTo(
+        new Double[]{60.0, 55.0, 50.0}));
+  }
+
+  @Test
+  public void shouldMergeTopKWithNulls() {
+    TopkKudaf<Double> doubleTopkKudaf = new TopkKudaf(0, 3, Double.class);
+    Double[] array1 = new Double[]{50.0, 45.0, null};
+    Double[] array2 = new Double[]{60.0, null, null};
```
Comment (satybald) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudafTest.java]: what about testing the case when value between nulls? i.e. `new Double[]{null, 60.0, null};`

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/StringTopkKudafTest.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+import static org.junit.Assert.assertThat;
+
+public class StringTopkKudafTest {
+
+  String[] valueArray;
+  @Before
+  public void setup() {
+    valueArray = new String[]{"10", "ab", "cde", "efg", "aa", "32", "why", "How are you", "Test",
+                              "123",
+                              "432"};
```
Comment (satybald) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/StringTopkKudafTest.java]: why new line needed for "423" value?

File: ksql-engine/src/main/java/io/confluent/ksql/util/ArrayUtil.java (hunk 9 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/util/ArrayUtil.java]: Good point! No need for toString. Removed it.

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudafTest.java (hunk 10 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/DoubleTopkKudafTest.java]: The merger will receive sorted arrays so we won't have a case where there are null values followed by not null values.

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/StringTopkKudafTest.java (hunk 11 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topk/StringTopkKudafTest.java]: IntelliJ misbehaved! Fixed it.

**Comment by hjafarpour** [comment]
@dguy applied your feedback.

File: ksql-engine/src/main/java/io/confluent/ksql/util/ArrayUtil.java (hunk 9 above)
Comment (satybald) [reply ksql-engine/src/main/java/io/confluent/ksql/util/ArrayUtil.java]: @hjafarpour while working on JSON_ARRAY_CONTAINS udf, I found a use case for converting to the string. When the parser converts json number as a long and search value is int conversion to string is needed. Example:

```
Object valueLong = (Object) 1L;
Object valueInt = (Object) 1;
valueLong.equals(valueInt); //false
```

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/IntTopkDistinctKudafTest.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  @Before
+  public void setup() {
+    valueArray = new Integer[]{10, 30, 45, 10, 50, 60, 20, 60, 80, 35, 25,
+                              60, 80};
+
+  }
+
+  @Test
+  public void shouldAggregateTopK() {
+    TopkDistinctKudaf<Integer> doubleTopkDistinctKudaf = new TopkDistinctKudaf(0, 3, Integer.class);
```
Comment (satybald) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/IntTopkDistinctKudafTest.java]: A variable name `doubleTopkDistinctKudaf` is misleading in all these cases.

File: ksql-engine/src/main/java/io/confluent/ksql/util/ArrayUtil.java (hunk 9 above)
Comment (satybald) [reply ksql-engine/src/main/java/io/confluent/ksql/util/ArrayUtil.java]: thought, if compare with toString() then it leads to unsound comparison logic.

```
Object valueString = (Object) "1";
Object valueInt = (Object) 1;
valueString.equals(valueInt); //false
valueString.toString().equals(valueInt.toString()); //true
```

File: ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/IntTopkDistinctKudafTest.java (hunk 12 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/IntTopkDistinctKudafTest.java]: Corrected the variable name.

**Pull request merged**